projects/rocr-runtime/runtime/hsa-runtime/image/hsa_amd_mipmap.cpp

```diff
@@ -0,0 +1,885 @@
+    }
+    return comps * comp_bits;
+}
+
+bool FillTileInfo(
+    hsa_agent_t agent,
+    const hsa_ext_image_descriptor_t* desc,
+    ADDR_TILEINFO& out_tile) {
+    (void)agent;
+    (void)desc;
+
+    std::memset(&out_tile, 0, sizeof(out_tile));
+    out_tile.banks = 4;
+    out_tile.bankWidth = 2;
+    out_tile.bankHeight = 2;
+    out_tile.macroAspectRatio = 1;
+    out_tile.tileSplitBytes = 0;
+    out_tile.pipeConfig = ADDR_PIPECFG_P2;
+    return true;
+}
+
+void PopulateAddrInput(
+    ADDR2_COMPUTE_SURFACE_INFO_INPUT* pIn,
+    const hsa_ext_image_descriptor_t* desc,
+    uint32_t num_mipmap_levels,
+    hsa_agent_t agent,
+    hsa_access_permission_t access_permission,
+    bool is_tiled) {
+    std::memset(pIn, 0, sizeof(*pIn));
+    pIn->size = sizeof(*pIn);
+
+    // Use dynamic surface flags instead of hardcoded values
+    pIn->flags = DetermineSurfaceFlags(desc, access_permission, is_tiled);
+
+    pIn->resourceType = GetAddrResourceType(desc->geometry);
+    pIn->format = HsaToAddrFormat(desc->format);
+
+    uint32_t bpp = ComputeBitsPerPixel(desc->format);
+    if (bpp == 0) {
+          bpp = 32;
+    }
+    pIn->bpp = bpp;
+
+    pIn->width = desc->width;
+    pIn->height = desc->height;
+    pIn->numSlices = (desc->array_size > 0) ? desc->array_size : 1;
+    pIn->numMipLevels = num_mipmap_levels;
+   static thread_local ADDR_TILEINFO tileInfo;
+    FillTileInfo(agent, desc, tileInfo);
+    pIn->pTileInfo = &tileInfo;
+
+    pIn->pitchInElement = 0;
+    pIn->sliceAlign = 0;
+    pIn->numSamples = 1;
+    pIn->numFrags = 0;
+
+    // Note: swizzleMode will be set by GetOptimalSwizzleMode after this function
+}
+
+#ifdef ADDR_GFX12_BUILD
+ADDR_E_RETURNCODE ComputeSurfaceInfoAddr3(
+    ADDR_HANDLE addr_lib,
+    hsa_agent_t agent,
+    const hsa_ext_image_descriptor_t* desc,
+    uint32_t num_mipmap_levels,
+    hsa_access_permission_t access_permission,
+    bool is_tiled,
+    ADDR3_COMPUTE_SURFACE_INFO_OUTPUT* addr_out) {
+
+    const core::Agent* agent_info = core::Agent::Convert(agent);
+    if (!agent_info) return ADDR_INVALIDPARAMS;
+
+    uint32_t gfx_major = agent_info->properties().major();
+    uint32_t gfx_minor = agent_info->properties().minor();
+
+    if (!UseAddr3ForArchitecture(gfx_major, gfx_minor)) {
+        return ADDR_INVALIDPARAMS; // Should use Addr2
+    }
+
+    ADDR3_COMPUTE_SURFACE_INFO_INPUT addr_in = {0};
+    addr_in.size = sizeof(addr_in);
+
+    addr_in.flags = DetermineSurfaceFlags(desc, access_permission, is_tiled);
+    addr_in.resourceType = GetAddrResourceType(desc->geometry);
+    addr_in.format = HsaToAddrFormat(desc->format);
+    addr_in.bpp = ComputeBitsPerPixel(desc->format);
+    if (addr_in.bpp == 0) addr_in.bpp = 32;
+
+    addr_in.width = desc->width;
+    addr_in.height = desc->height;
+    addr_in.numSlices = (desc->array_size > 0) ? desc->array_size : 1;
+    addr_in.numMipLevels = num_mipmap_levels;
+    addr_in.numSamples = 1;
+    addr_in.numFrags = 0;
```

Comment:
Addrlib clamps this so it has no effect, but technically numFrags should also be 1.